### ISSUE TITLE
jsonwrite for outdated SPM

### ISSUE THREAD

**Issue by nbeliy** [bids-standard/bids-matlab#176]
The interface of `spm_jsonwrite` changed in SPM in revision r. 6975, making it unusable for outdated SPM installations. 
Installation of JSONio do not change anything as the check for existence for `spm_jsonwrite` comes first.

I see two solution for this (other from keep SPM updated):

- use only JSONio, cite it as dependency
- Invert if in `+bids/+util/jsonencode.m` line 33, so JSONio takes priority

Tell me that you think

UPD:
You also check for native MATLAB jsonencode, but it has the lower priority. May be it can be rendered as highest one

**Comment by Remi-Gau** [comment]
Thanks for spotting that.

For the moment, I am in favor of reordering the priorities: because adding "hard dependencies" with matlab can be annoying due to the lack of package management: unless we decide to ship it with bids-matlab.

For priorities: how about JSONio > SPM > Matlab?

I thought of putting Matlab first but then realized that the `json_encode` only appeared in Matlab 2016, so I currently ordered them by how "heavy" those "dependencies" are.


Personnal note: how is the JSON support on octave 5.1 these days?

**Comment by nbeliy** [comment]
Naively I would prefer first Matlab: Matlab is guaranteed to be maintained with same interface. 
Then JSONio: independent package not linked with organisation -- danger to disappear.
FInally SPM, with danger of interface change.

But I'm only starting to discover all the versions dependencies, so take my opinion at low weight

**Comment by Remi-Gau** [comment]
Yeah that's a good point as well.

Will ping @gllmflndn for this take as he knows a thing or 2 about JSONio and SPM. :wink:

**Comment by gllmflndn** [comment]
We should rely on native implementations (`jsonencode`/`jsondecode`) first as everything else is just a temporary solution that will become deprecated as soon as native options are widespread.

`spm_BIDS` (`JSONio`, `bids-matlab`) predates `jsonencode`/`jsondecode` that were introduced in MATLAB R2016b and improved in R2017b (`ConvertInfAndNaN`) and R2021a (`PrettyPrint`). These functions are now implemented in the development version of Octave and will be available in Octave 7.1 that, I guess, won't be released until 2022.

Given this timeline, it is still relevant to have workarounds in place in the absence of built-in JSON functionality. I am not sure whether to check for SPM or JSONio first matters much. I guess the message to someone wanting to use `bids-matlab` with a incompatible version of MATLAB or Octave is to install `JSONio` unless they are SPM users where they can rely on `spm_jsonread`/`spm_jsonwrite`. `JSONio` is just a spin-off the SPM functions, to make them more acceptable as a dependency.

Concerning the change in interface, JSON functions were added during SPM12's lifetime so, sure, the first version of SPM12 did not have them. The syntax might have changed to add new features (r6975 you mention is already 5 years old) or match MATLAB when the functions were introduced there but note the same is happening with these functions (you can only rely on `PrettyPrint` from R2021a). As for the comment about 'danger to disappear', I will be careful next time I cross the street.

In short: 👍 for using `jsonencode`/`jsondecode` first and foremost and let's remove everything else in 2023?

**Comment by Remi-Gau** [comment]
Thanks for all the clarifications.

What you outline matches what I did in the PR #178 so I will merge the change.

> As for the comment about 'danger to disappear', I will be careful next time I cross the street.

Don't worry, we are not about to send people after you. We love you too much!!! :heart:
